### DISCUSSION TITLE
Possible to use as a server-side API?

### DISCUSSION THREAD

**Discussion by adamquaile** [alpinejs/alpine#596] [4. General]
This might be out of scope for the project, but I'll ask anyway. 

Would it be possible to use this library entirely on the server side, for prerendering content? 

E.g. 

```js
import { render } from 'alpinejs'
html = render('<html>...<span v-text="var">...');
```

For more context, I like how small the learning curve is on alpine, and I was actually thinking this could make a very nice thin wrapper around HTML, or even MJML to build emails on. 

In this case an API might be required to pass extra parameters, like in traditional template languages for some of the variables. 

```js
import { render } from 'alpinejs'
html = render('<html>...<span v-text="var">...', { var: 'a server-side variable' });
```

It seems like this might be hard if no DOM actually exists on the server side, but is this possible or even a good idea? 

Thanks!

**Comment by HugoDF** [comment]
It's totally out of the scope of that project, but you can use the `render()` util from https://github.com/HugoDF/alpine-test-utils and get the `outerHTML` to do what you wish:

```js
import {render} from 'alpine-test-utils'

html = render('<div x-data="{ aa: 'foo' }"></div>', { aa: 'bar' }).outerHTML;
```

> Note: in the example above, `aa` will be overridden to 'bar'

See https://github.com/HugoDF/alpine-test-utils#quick-start-write-your-first-test

**Reply by adamquaile** [reply]
That looks like a good option. I had thought of using a headless browser, but this looks better since it uses jsdom. 

Thanks for the pointer, I'll give it a try.

**Reply by HugoDF** [reply]
One thing to note, x-text doesn't work properly in JSDOM since it sets innerText and JSDOM doesn't implement it

**Reply by adamquaile** [reply]
Ah, that would explain why it didn't work when I tried it. That's a bit of a deal-breaker for this particular idea. No worries, I'll either try with a headless browser or give up, it's not important.

Thanks for the help!

**Reply by ryangjchandler** [reply]
Is there any benefits to using Alpine over regular template literals on the server, since you don't really need the reactivity.

I'd prefer to use:

```js
let html = `
    <p>${text}</p>
`
```

Than run things through a JSDOM implementation..

**Reply by adamquaile** [reply]
The benefits I see would be in looping and conditional constructs which could get cumbersome with native string manipulation as in your example. 

So, essentially it would be taking the template parts of alpine and ignoring the reactivity parts so I totally understand why it's out of scope, but I can imagine use-cases for it, and it's a nice, low learning curve template language. 

I can use handlebars or twig, or any number of template languages instead.

**Reply by adamquaile** [reply]
Also, beyond my own use case, I don't know if it could be applicable at all for server-side rendering in a JS framework, but there might be some use cases there where a server-side api would help 🤷

**Reply by ryangjchandler** [reply]
Fair enough. With template literals, most of the stuff can be done in a React fashion. Loops can be done with a `map()` and a final `join()`, conditionals can be done with a ternary or similar. Either way, it would be interesting to see Alpine used in that scenario.
